src/primaite/agents/aegis/gllm.py

```diff
@@ -89,10 +96,62 @@ def get_cosine_embeddings_loss(self, openai_hidden_states, gllm_hidden_states):
 
         return loss
 
+    def get_crossentropy_loss(self, gt_probs, gt_tokens, gllm_probs, gllm_tokens):
+        criterion = torch.nn.CrossEntropyLoss(reduction="none")
+        padding_logit = torch.zeros([len(self.llm.tokenizer)])
+        padding_logit[self.llm.tokenizer.pad_token_id] = (
+            100  # Mock pad token of gt to 100% prob (it gets ignored anyway because of the mask)
+        )
+        padding_token = torch.tensor([self.llm.tokenizer.pad_token_id])  # Set padding token tensor
+
+        # Pad the shortest set of tensors if required
+        batch_size = gllm_probs.shape[0]
+        if gt_probs.shape[1] > gllm_probs.shape[1]:
+            # gt sequences are longer - pad gllm tokens
+            n_pads = gt_probs.shape[1] - gllm_probs.shape[1]
+            gt_probs = gt_probs[:, : gllm_probs.shape[1], :]
+            gt_tokens = gt_tokens[:, : gllm_tokens.shape[1]]
+
+        elif gllm_probs.shape[1] > gt_probs.shape[1]:
+            # gllm sequences are longer - pad openai responses
+            n_pads = gllm_probs.shape[1] - gt_probs.shape[1]
+            padding_logits = padding_logit.repeat(batch_size, n_pads, 1)
+
+            gt_probs = torch.cat([gt_probs, padding_logits], dim=1)
+
+            padding_tokens = padding_token.repeat(batch_size, n_pads)
+            gt_tokens = torch.cat([gt_tokens, padding_tokens], dim=1)
+
+        # Calculate the cross entropy loss - transpose the 1 and 2 dimensions because we want a loss value per token not per vocab
+        loss = criterion(gllm_probs.transpose(1, 2), gt_probs.transpose(1, 2))
+        loss = loss.mean()
+
+        return loss
+
+
+def get_n_trainable_llm_parameters(model):
+    trainable_model_params = 0
+    all_model_params = 0
+    for _, param in model.named_parameters():
+        all_model_params += param.numel()
+        if param.requires_grad:
+            trainable_model_params += param.numel()
+    return f"Trainable LLM parameters: {trainable_model_params}\nAll LLM parameters: {all_model_params}\nPercentage of trainable LLM parameters: {100 * trainable_model_params / all_model_params:.2f}%"
```

Comment:
I am a Python nerd :) so please make this f-string nice lol 
If you add like `{trainable_model_params:_}` it will print large numbers like 9_000_000
And for percentages you can just `{trainable_model_params / all_model_params:.2%}`

Reply:
Mate take this up with deeplearning.ai cos I robbed this code from their course generative ai with LLMs xD

Reply:
Shocking from them

Reply:
bruvvah I didn't even know about :_ or :, wowowwww